gocd/gocd.go

```diff
@@ -286,6 +286,124 @@ func (server Server) pipelineGET(pipelineName string) (pipeline Pipeline, etag s
 	return
 }
 
+// Partially generated by curl-to-Go: https://mholt.github.io/curl-to-go
+func (server Server) artifactGET(pipelineName string, pipelineID int, stageName string, stageID int, jobName string, artifactPath string) (fileBytes *bytes.Buffer, err error) {
+	reqStr := fmt.Sprintf("%s/go/files/%s/%d/%s/%d/%s/%s/", server.URL(),
+		pipelineName, pipelineID, stageName, stageID, jobName, artifactPath)
+	req, err := http.NewRequest("GET", reqStr, nil)
+	if err != nil {
+		return
+	}
+
+	if len(server.User) > 0 && len(server.Password) > 0 {
+		req.SetBasicAuth(server.User, server.Password)
+	}
+	req.Header.Set("Content-Type", "application/json")
+
+	log.Debugf("Sending request: %v", req)
+	resp, err := client(server.Timeout).Do(req)
+	if err != nil {
+		return
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return
+	}
+
+	if resp.StatusCode != 200 {
+		err = fmt.Errorf("Bad response code: %d with response: %s", resp.StatusCode, body)
+		return
+	}
+
+	fileBytes = bytes.NewBuffer(body)
+	return
+}
+
+// Partially generated by curl-to-Go: https://mholt.github.io/curl-to-go
+func (server Server) historyGET(pipelineName string) (historyJSON bytes.Buffer, err error) {
+	req, err := http.NewRequest("GET", server.URL()+"/go/api/pipelines/"+pipelineName+"/history", nil)
+	if err != nil {
+		return
+	}
+
+	if len(server.User) > 0 && len(server.Password) > 0 {
+		req.SetBasicAuth(server.User, server.Password)
+	}
+	req.Header.Set("Content-Type", "application/json")
+
+	log.Debugf("Sending request: %v", req)
+	resp, err := client(server.Timeout).Do(req)
+	if err != nil {
+		return
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return
+	}
+
+	if resp.StatusCode != 200 {
+		err = fmt.Errorf("Bad response code: %d with response: %s", resp.StatusCode, body)
+		return
+	}
+
+	err = json.Indent(&historyJSON, body, "", "\t")
+	if err != nil {
+		log.Warn("Failed to prettify JSON: ", err)
+	}
+
+	// log.Debug("pipelineHistory JSON:", string(prettyJSON.Bytes()))
+	return
+}
+
+// History gets the run history of a pipeline of a given name exist. returns map
+func History(server *Server, name string) (latestRuns map[string]int, err error) {
```

Comment:
Let's add some more comments to better annotate what's going on (for others and our future selves)